index.d.ts

```diff
@@ -1,19 +1,66 @@
+declare namespace unusedFilename {
+	interface Options {
+		/**
+		Custom function to increment a filename. By default `parenthesesIncrementer` is used.
+		*/
+		incrementer?: Incrementer;
+
+		/**
+		Max number of attempts to find an unused filename before giving up and returning the last tried name.
+		@default Infinity
+		*/
+		maxTries?: number;
```

Comment:
Not sure how useful this option is. It can cause bad problems if the user sets this, it get's hit, they get a used filename back and then overwrites an important file.

I see 2 other options:

1. Throw a custom error the user can check for and handle the case properly.
2. Just append a random number ID/number to the file if it's hit.

Reply:
`maxTries` is one of the reasons I made this PR since I need it. My use case is running an operation that does something to the original file, and generates a new one next to it, but the original should always be kept, while the new one replaced in case the operation is run on the original again. In this case I'm setting `maxTries` to `1`.

In my head at least, the potential for overwriting an existing file is an expected behavior when setting this option to anything below `Infinity`.

If developer sets this to a finite number, and at the same time they want to ensure no overwrites (which is weird, since these seem like mutually exclusive intents imo), they should add a second file exists check after the return.

Thoughts?

Reply:
> In my head at least, the potential for overwriting an existing file is an expected behavior when setting this option to anything below Infinity.

It's expected if you read the docs carefully, yes. But not all users do that.

> Imo the whole point of maxTries is to specify a point after which you don't care if it's going to overwrite.

That is your use-case. Actually, the main use-case I can think of is to prevent running for too long or abuse. For example, a consumer might want to set it to 100, which will probably never be hit while still ensuring that it will not run for a very long time.

I have personally implemented such a logic in my macOS apps where I just set a hard-coded limit to 1000, afterwhich I just append a UUID.

Reply:
If it just returns the same name as previous and you need to handle it differently, it becomes a bit difficult to handle.

I think throwing a custom error would solve both your use-case and any other, as that makes it up to the consumer to decide what to do. No assumptions.

Reply:
You're right, the main point of this module is to not overwrite files, so one should expect reaching max tries would throw. Was blinded by my requirements :)

What should the custom error look like? Is there some precedent error code in node similar to this?

I guess we should also add the last tested path to it?

It'd also be useful to add the original path (stripped of the incrementer pre/suf-fix) to it, but in that case the incrementer would have to return a `[originalPath, incrementedPath]` tuple, so that we know what that is.

Reply:
Just `export class MaxTryError extends Error {}` with some added metadata like you mentioned.

Reply:
`MaxTryError` makes it sound like it throws always when max tries is reached, but it will only throw when the file still doesn't exist. Shouldn't it be something like `NotFoundError`?

Reply:
> MaxTryError makes it sound like it throws always when max tries is reached

That's how it should work. Sorry if I didn't make that clear enough. It's up to the user how they want to handle it.